sync/syncer.go

```diff
@@ -201,29 +238,71 @@ func (s *Syncer) Synchronise() {
 	currentSyncLayer := s.lValidator.ValidatedLayer() + 1
 	if s.lastTickedLayer() == 1 { // skip validation for first layer
 		s.Info("Not syncing in layer 1")
+		s.setP2pSynced(true)
 		return
 	}
-	if s.IsSynced() {
-		lyr, err := s.GetLayer(types.LayerID(currentSyncLayer))
+	if s.WeaklySynced() {
+		s.Info("Node is synced. Going to validate layer %v", currentSyncLayer)
+
+		lyr, err := s.lProvider.GetLayer(currentSyncLayer)
 		if err != nil {
-			s.With().Error("failed getting layer even though IsSynced is true", log.Err(err))
+			s.With().Error("failed getting layer even though we are weakly-synced", log.Err(err))
 			return
 		}
 		s.lValidator.ValidateLayer(lyr) // wait for layer validation
 		return
 	}
-	for ; currentSyncLayer <= s.lastTickedLayer(); currentSyncLayer++ {
-		lastTickedLayer := s.lastTickedLayer()
-		s.Info("syncing layer %v current consensus layer is %d", currentSyncLayer, lastTickedLayer)
+
+	s.Info("Node is out of sync setting p2p-synced to false and starting sync")
+	s.setP2pSynced(false)
+
+	for ; currentSyncLayer < s.lastTickedLayer(); currentSyncLayer++ {
+		s.Info("syncing layer %v current consensus layer is %d", currentSyncLayer, s.lastTickedLayer())
 		lyr, err := s.getLayerFromNeighbors(currentSyncLayer)
 		if err != nil {
 			s.Info("could not get layer %v from neighbors %v", currentSyncLayer, err)
 			return
 		}
-		if currentSyncLayer < lastTickedLayer {
-			s.lValidator.ValidateLayer(lyr) // wait for layer validation
-		}
+
+		s.lValidator.ValidateLayer(lyr) // wait for layer validation
+	}
+
+	_, err := s.getLayerFromNeighbors(currentSyncLayer)
+	if err != nil {
+		s.Info("could not get last ticked layer %v from neighbors %v", currentSyncLayer, err)
+		return
+	}
+
+	s.p2pSyncForTwoLayers(currentSyncLayer)
+}
+
+func (s *Syncer) p2pSyncForTwoLayers(currentSyncLayer types.LayerID) {
+	// subscribe and wait for two ticks
+	ch := s.clockSub.Subscribe()
+	<-ch
+	<-ch
+	s.clockSub.Unsubscribe(ch) // unsub, we won't be listening on this ch anymore
+
+	// assumed to be weakly synced here
+	// just get the layers and validate
+
+	lyr, err := s.lProvider.GetLayer(currentSyncLayer)
+	if err != nil {
```

Comment:
this function can return errors and have them printed in calling function

Reply:
Done